internal/loader/configuration_client_manager.go

```diff
@@ -0,0 +1,478 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package loader
+
+import (
+	"azappconfig/provider/internal/properties"
+	"context"
+	"fmt"
+	"math"
+	"net"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	acpv1 "azappconfig/provider/api/v1"
+
+	"math/rand"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/data/azappconfig"
+	"github.com/google/uuid"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+//go:generate mockgen -destination=mocks/mock_configuration_client_manager.go -package mocks . ClientManager
+
+type ConfigurationClientManager struct {
+	ReplicaDiscoveryEnabled   bool
+	SpecifiedClient           *azappconfig.Client
+	DynamicClients            []*ConfigurationClientWrapper
+	ValidDomain               string
+	Endpoint                  string
+	ClientsFallbackStatus     map[string]*ConfigurationClientWrapper
+	Credential                azcore.TokenCredential
+	secret                    string
+	id                        string
+	lastFallbackClientAttempt metav1.Time
+	lastFallbackClientRefresh metav1.Time
+}
+
+type ConfigurationClientWrapper struct {
+	Endpoint       string
+	Client         *azappconfig.Client
+	BackOffEndTime metav1.Time
+	FailedAttempts int
+}
+
+type ClientManager interface {
+	GetClients() []*ConfigurationClientWrapper
+	RefreshClients()
+	UpdateClientBackoffStatus(endpoint string, successful bool)
+	ExecuteFailoverPolicy(ctx context.Context, filters []acpv1.Selector, getSettingsFunc GetSettingsFunc) ([]azappconfig.Setting, error)
```

Comment:
I don't expect the clientManager has the method `ExecuteFailoverPolicy()`, the clientManger just take the responsibility for managing the clients.

Reply:
Agree, will move it to configLoader.